main/ertp/api/contract-hosts.md

```diff
@@ -0,0 +1,158 @@
+# Contract Hosts
+
+ContractHost provides a way to install and run verifiable contracts.
+
+Install(source) creates Installations. Installations represent particular
+contract forms, which can be spawn()'d to create Contract instances
+initialized with specific terms. These Contracts manage invitations for
+multiple seats, each of which repesents a "role" in an interaction.
+
+Some seats provide the methods `getWinnings()` and `getRefunds()`, which return
+promises for payments for the outputs of the interaction. We provide the
+`collect()` method (described later) to simplify collection of winnings and
+refunds into appropriate purses.
+
+## contractHost.install(contractSrc)
+- `contractSrc` `{String}` - contractSrc is the source code for a Contract. See Contract for details.
+- Returns: `{Installation}`
+
+Install the source code for a Contract.
+
+```js
+const someEscrowInstallation = E(someHost).install(someEscrowExchangeSrc);
+```
+
+## contractHost.getInstallationSourceCode(installation)
+- `installation` `{Installation}`
+- Returns: `{String}`
+
+Get the actual source code of the Installation.
+
+```js
+const theSourceCode = E(someHost).getInstallationSourceCode(someInstallation);
+```
+
+## contractHost.redeem(invite)
+- `invite` `{Payment}`
+- Returns: `{Object}`
+
+Seat invitations are packaged as payments from the Invite Assay. Redeeming an invite returns a seat object with an arbitrary interface (the interface is at the discretion of the contract) which supports interaction according to the terms.
+
+```js
+const someSeat = E(someHost).redeem(someInvite);
+```
+
+## contractHost.getInviteAssay()
+- Returns: `{Assay}`
+
+The assay allows holders of seat invitations to get exclusive access to a Seat.
+
+```js
+const assayWithExclusiveAccess = E(someHost).getInviteAssay();
+```
+
+## installation.spawn(terms)
+- `terms` `{Terms}`
+- Returns: `{Invites <Object>}`
+
+Create a new `InviteMaker`, then call the Contract's `start()` method and return its results. The results are often a collection of seat invitations for the seats in the contract, but see [`coveredCall`](https://github.com/Agoric/ERTP/blob/master/core/coveredCall.js) for an exception.
+
+```js
+import harden from '@agoric/harden';
+
+const terms = harden({ left: whatPartyAWants, right: whatPartyBWants });
+const invitePayments = E(someInstallation).spawn(terms);
+```
+
+## installation.checkUnits(installation, inviteUnits, terms)
+- `installation` `{Installation}`
+- `inviteUnits` `{Units}`
+- `terms` `{Terms}`
+- Returns:
+
+### <span style="color:red">What does this method return? It's not stated in the chainmail file and it's not very clear in the code usage.</span>
+
+### <span style="color:red">Are these params correct? In the code they follow this pattern instead: `{inviteUnits, terms, anOptionalString}`. Example from `ERTP/test/swingsetTests/contractHost/vat-alice.js`:</span>
```

Comment:
I think these arguments are good. In this case, I think { left, right} are the terms so this matches the pattern

Reply:
I meant that the chainmail file suggests that the params should be 
`(installation, inviteUnits, terms)`, 
whereas in the code I see a different order/usage of params that look more like
`(inviteUnits, terms, anOptionalString)`.
So it's like the `installation` param got removed and an option String param got added to the  end. So that's what I'm confused about.

Reply:
Ah gotcha. I'm not sure about which is right. Do you know @Chris-Hibbert

Reply:
Yeah, this is kind of weird, and is another place like the contract seat, where the developer can decide on different parameters for different contract, and different method names. Some contracts will have more than one check...() method that are called differently. [Escrow has checkUnits() and checkPartialUnits()].

The convention I wanted to establish was that the methods would usually be written with their first two parameters being the units and the terms. Other parameters can be added. Escrow adds the seat as an extra parameter, but that's not necessary when there's no distinction between possible roles. (All clients of an auction get the same kind of seat.)

The check function as written in the contract file, has installation, units, and terms. The user who has access to a seat from an invite and wants to validate it, calls it with all but the first parameter, which is curried in by the contract host when it evaluates the check...() function. So the version to be called is different from the version to be written, and I wasn't clear about that.

Reply:
Thanks for the clarification! I updated some of the wording and added emphasis on parameter order and the difference between calling the functions on the dev side vs user side.